chapter_recurrent-neural-networks/text-preprocessing.ipynb

```diff
@@ -0,0 +1,370 @@
+    "Each text sequence is split into a list of tokens.\n",
+    "A *token* is the basic unit in text.\n",
+    "In the end,\n",
+    "a list of token lists are returned,\n",
+    "where each token is a string.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public String[][] tokenize(String[] lines, String token) throws Exception {\n",
+    "    // Split text lines into word or character tokens.\n",
+    "    String[][] output = new String[lines.length][];\n",
+    "    if (token == \"word\") {\n",
+    "        for (int i = 0; i < output.length; i++) {\n",
+    "            output[i] = lines[i].split(\" \");\n",
+    "        }\n",
+    "    }else if (token == \"char\") {\n",
+    "        for (int i = 0; i < output.length; i++) {\n",
+    "            output[i] = lines[i].split(\"\");\n",
+    "        }\n",
+    "    }else {\n",
+    "        throw new Exception(\"ERROR: unknown token type: \" + token);\n",
+    "    }\n",
+    "    return output; \n",
+    "}\n",
+    "String[][] tokens = tokenize(lines, \"word\");\n",
+    "for (int i = 0; i < 11; i ++) {\n",
+    "    System.out.println(Arrays.toString(tokens[i]));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 8
+   },
+   "source": [
+    "## Vocabulary\n",
+    "\n",
+    "The string type of the token is inconvenient to be used by models, which take numerical inputs.\n",
+    "Now let us build a dictionary (HashMap), often called *vocabulary* as well, to map string tokens into numerical indices starting from 0.\n",
+    "To do so, we first count the unique tokens in all the documents from the training set,\n",
+    "namely a *corpus*,\n",
+    "and then assign a numerical index to each unique token according to its frequency.\n",
+    "Rarely appeared tokens are often removed to reduce the complexity.\n",
+    "Any token that does not exist in the corpus or has been removed is mapped into a special unknown token “&lt;unk&gt;”.\n",
+    "We optionally add a list of reserved tokens, such as\n",
+    "“&lt;pad&gt;” for padding,\n",
+    "“&lt;bos&gt;” to present the beginning for a sequence, and “&lt;eos&gt;” for the end of a sequence.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public class Vocab {\n",
+    "    public int unk;\n",
+    "    public List<Map.Entry<String, Integer>> tokenFreqs;\n",
+    "    public List<String> idxToToken;\n",
+    "    public HashMap<String, Integer> tokenToIdx;\n",
+    "\n",
+    "    public Vocab(String[][] tokens, int minFreq, String[] reservedTokens) {\n",
+    "        // Sort according to frequencies\n",
+    "        HashMap<String, Integer> counter = countCorpus2D(tokens);\n",
+    "        this.tokenFreqs = new LinkedList<Map.Entry<String, Integer>>(counter.entrySet()); \n",
```

Comment:
any reason why using LinkedList rather than ArrayList here?

Reply:
Nope. The only time I tried using this type of sort method with comparator, I followed a tutorial using LinkedList and I forgot to change it to ArrayList after testing. Just changed it. Thanks for noticing.